core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -71,79 +75,211 @@ object TransactionMarkerChannelManager {
       false,
       new ApiVersions
     )
-    val channel = new TransactionMarkerChannel(config.interBrokerListenerName, metadataCache, networkClient, time)
-
-    val sendThread: InterBrokerSendThread = {
-      networkClient.wakeup()
-      new InterBrokerSendThread(threadName, networkClient, requestGenerator(channel, txnMarkerPurgatory), time)
-    }
 
     new TransactionMarkerChannelManager(config,
       metadataCache,
+      networkClient,
+      txnStateManager,
       txnMarkerPurgatory,
-      sendThread,
-      channel)
+      time)
   }
 
-  private[transaction] def requestGenerator(transactionMarkerChannel: TransactionMarkerChannel,
-                                            txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]): () => Iterable[RequestAndCompletionHandler] = {
-    () => transactionMarkerChannel.drainQueuedTransactionMarkers(txnMarkerPurgatory)
+  private[transaction] def requestGenerator(transactionMarkerChannelManager: TransactionMarkerChannelManager): () => Iterable[RequestAndCompletionHandler] = {
+    () => transactionMarkerChannelManager.drainQueuedTransactionMarkers()
   }
 }
 
+class TxnMarkerQueue(@volatile private var destination: Node) {
+
+  // keep track of the requests per txn topic partition so we can easily clear the queue
+  // during partition emigration
+  private val markersPerTxnTopicPartition: concurrent.Map[Int, BlockingQueue[TxnIdAndMarkerEntry]]
+  = concurrent.TrieMap.empty[Int, BlockingQueue[TxnIdAndMarkerEntry]]
+
+  def removeMarkersForTxnTopicPartition(partition: Int): Option[BlockingQueue[TxnIdAndMarkerEntry]] = {
+    markersPerTxnTopicPartition.remove(partition)
+  }
+
+  def maybeUpdateNode(node: Node): Unit = {
+    destination = node
+  }
+
+  def addMarkers(txnTopicPartition: Int, txnIdAndMarker: TxnIdAndMarkerEntry): Unit = {
+    val queue = markersPerTxnTopicPartition.getOrElseUpdate(txnTopicPartition, new LinkedBlockingQueue[TxnIdAndMarkerEntry]())
+    queue.add(txnIdAndMarker)
+  }
+
+  def forEachTxnTopicPartition[B](f:(Int, BlockingQueue[TxnIdAndMarkerEntry]) => B): mutable.Iterable[B] =
+    markersPerTxnTopicPartition.filter { case(_, queue) => !queue.isEmpty }
+      .map { case(partition:Int, queue:BlockingQueue[TxnIdAndMarkerEntry]) => f(partition, queue) }
 
+  def node: Node = destination
+
+  // TODO: this function is only for metrics recording, not yet added
+  def totalNumMarkers(): Int = markersPerTxnTopicPartition.map { case(_, queue) => queue.size()}.sum
+}
 
 class TransactionMarkerChannelManager(config: KafkaConfig,
```

Comment:
After we remove `TransactionMarkerChannel`, maybe we can claim its name for this class?